wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala

```diff
@@ -0,0 +1,283 @@
+              .sequence(txoProcessingFutures)
+              .map(_.flatten)
+
+        }
+
+    val outgoingTxFut: Future[Vector[IncomingWalletTXO]] = {
+      for {
+        outputsBeingSpent <- incomingTxoDAO.findOutputsBeingSpent(transaction)
+        processed <- FutureUtil.sequentially(outputsBeingSpent)(
+          markAsSpentIfUnspent)
+      } yield processed.flatten.toVector
+
+    }
+
+    val aggregateFut =
+      for {
+        incoming <- incomingTxoFut
+        outgoing <- outgoingTxFut
+      } yield {
+        ProcessTxResult(incoming.toList, outgoing.toList)
+      }
+
+    aggregateFut.failed.foreach { err =>
+      val msg = s"Error when processing transaction=${transaction.txIdBE}"
+      logger.error(msg, err)
+    }
+
+    aggregateFut
+  }
+
+  /** If the given UTXO is marked as unspent, updates
+    * its spending status. Otherwise returns `None`.
+    */
+  private val markAsSpentIfUnspent: (
+      (SpendingInfoDb, IncomingWalletTXO)) => Future[
+    Option[IncomingWalletTXO]] = {
+    case (info, out) =>
+      if (out.spent) {
+        Future.successful(None)
+      } else {
+        val updatedF =
+          incomingTxoDAO.update(out.copy(spent = true))
+        updatedF.foreach(
+          updated =>
+            logger.debug(
+              s"Marked utxo=${info.toHumanReadableString} as spent=${updated.spent}")
+        )
+        updatedF.map(Some(_))
+      }
+  }
+
+  /**
+    * Inserts the UTXO at the given index into our DB, swallowing the
+    * error if any (this is because we're operating on data we've
+    * already verified).
+    */
+  private def processUtxo(
+      transaction: Transaction,
+      index: Int): Future[SpendingInfoDb] =
+    addUtxo(transaction, UInt32(index))
+      .flatMap {
+        case AddUtxoSuccess(utxo) => Future.successful(utxo)
+        case err: AddUtxoError =>
+          logger.error(s"Could not add UTXO", err)
+          Future.failed(err)
+      }
+
+  private case class OutputWithIndex(output: TransactionOutput, index: Int)
+
+  /**
+    * @param xs UTXO sequence we want to create
+    *           incoming wallet TXOs for
+    */
+  private def createIncomingTxos(
+      transaction: Transaction,
+      confirmations: Int,
+      xs: Seq[(SpendingInfoDb, OutputWithIndex)]): Vector[IncomingWalletTXO] = {
+    xs.map {
+      case (utxo, OutputWithIndex(out, _)) =>
+        IncomingWalletTXO(
+          confirmations = confirmations,
+          // is this always the case?
+          spent = false,
+          scriptPubKey = out.scriptPubKey,
+          txid = transaction.txIdBE,
+          // always defined, as its freshly
+          // written to the DB
+          spendingInfoID = utxo.id.get
+        )
+    }.toVector
+  }
+
+  /**
+    * Processes an incoming transaction that already exists in our wallet.
+    * If the incoming transaction has more confirmations than what we
+    * have in the DB, we update the TX
+    */
+  private def processExistingIncomingTxo(
```

Comment:
I don't like the return type here, let's keep it `Future[IncomingWalletTxo]` and explicitly throw in the case of reorgs. Currently we call `.flatten` inside of `processTransactionImpl` which silently disregards the `Option` type returned here.